File: src/main/java/me/sullivan/glasslang/parser/nodes/FunctionDefinitionNode.java
package me.sullivan.glasslang.parser.nodes;

import me.sullivan.glasslang.interpreter.Interpreter;
import me.sullivan.glasslang.interpreter.primitives.types.Primitive;
import me.sullivan.glasslang.lexer.token.Token;

import java.util.List;

public class FunctionDefinitionNode extends Node
{

    private final List<Token> args;
    private final Node body;

    public FunctionDefinitionNode(Token token, List<Token> args, Node body)
    {
        super(token);
        this.args = args;
        this.body = body;
    }

    @Override
    public Object getValue()
    {
        return token.getValue();
    }

    @Override
    public String toString()
    {
        return "Function(" + (token != null ? token.getValue() : "anonymous-func") + ", " + args + " => " + body;
    }

    @Override
    public Primitive<?> visitor(Interpreter interpreter)
    {
        return interpreter.visitFuncDefinitionNode(this);
    }

    public List<Token> getArgs()
    {
        return args;
    }

    public Node getBody()
    {
        return body;
    }
}

File: src/me/sullivan/glasslang/parser/nodes/FunctionDefinitionNode.kt
package me.sullivan.glasslang.parser.nodes

import me.sullivan.glasslang.interpreter.Interpreter
import me.sullivan.glasslang.interpreter.primitives.Primitive
import me.sullivan.glasslang.lexer.token.Token

class FunctionDefinitionNode(token: Token, val args: List<Token>, val body: Node) : Node(token)
{
    override val value: Any
        get() = token.value

    override fun toString(): String
    {
        val name = if (token.value != null) token.value else "anonymous-func"
        return "Function($name) , $args => $body"
    }

    override fun visitor(interpreter: Interpreter): Primitive<*>
    {
        return interpreter.visitFuncDefinitionNode(this)
    }
}